src/model_DAforcing.py

```diff
@@ -123,6 +126,201 @@ def __init__(self, bmi_cfg_file=None):
             if lastobs_file:
                 self._lastobs_df = _read_lastobs_file(lastobs_file)
 
+            # save time reference: start_datetime
+            self.dateNull = start_datetime
+
+            # read in metadata for BMI compliant arrays:
+
+            # USGS Observations
+            if nudging or usgs_persistence:
+
+                # 
+                # USGS dataframe: 
+                # 
+                # Dates related:
+                # - datesSecondsArray_usgs: "column names" of usgs dataframe: dates
+                #        in seconds relative to start_datetime
+                # - nDates_usgs: number of "dates" columns
+                # 
+                # Station ID related:
+                # - stationArray_usgs: "row names" of usgs dataframe: all station
+                #        names decoded into arrays of one ASCII code per string,
+                #        attached together
+                # - stationStringLengthArray_usgs: array of lengths of station IDs,
+                #        in case of mixed arrays (not all station IDs have equal lengths)
+                # - nStations_usgs: number of stations
+                #
+                ( _datesSecondsArray_usgs, _nDates_usgs, _stationArray_usgs, \
+                    _stationStringLengthArray_usgs, _nStations_usgs) \
+                    = _time_stations_from_df(self._usgs_df,start_datetime)
+                # save metadata in class instances
+                self.datesSecondsArray_usgs = _datesSecondsArray_usgs
+                self.nDates_usgs = _nDates_usgs
+                self.stationArray_usgs = _stationArray_usgs
+                self.stationStringLengthArray_usgs = _stationStringLengthArray_usgs
+                self.nStations_usgs = _nStations_usgs
+                # flatten the actual USGS datafrane into a numpy ndarray
+                _usgsArray = _flatten_array(self._usgs_df)
+                # ... and save it with the class instance
+                self.usgsArray = _usgsArray
+
+                # 
+                # Reservoir USGS and USACE dataframes: Array structures of 
+                # _reservoir_usgs_df and _reservoir_usace_df are
+                # fundamentally the same as usgs_df: input/output analogous.
+                # 
+                # In the following comments, usgs shall also represent usace
+                #
+                # Dates related:
+                # - datesSecondsArray_reservoir_usgs
+                # - nDates_usgs
+                # 
+                # Station ID related:
+                # - stationArray_usgs 
+                # - stationStringLengthArray_usgs 
+                # - nStations_usgs 
+                #
```

Comment:
This class seems to be getting a lot of attributes that are grouped/related. I strongly recommend creating new data structures/classes for these groups and then holding an instance in the class. This could also provide the basis for an abstract interface that allows new DA sources to more easily plug in in the future.

Reply:
That point is well taken, although the treatment of the dataframes is not as "universal" as it may appear at first. For example, the usgs- and usace dataframes (and, by extension, any future additional gauges included) are indeed similar (once preprocessed), but the rfc arrays, for example,, are quite different. 

However, once all the dataframes have been outfitted with a BMI transport option (a few weeks at most), we'll definitely evaluate opportunities to create new classes. One candidate is definitely said preprocessed usgs- and usace arrays (vs station IDs and date/time).